File: Python/Linked List/Remove The Nth Node From The End.py
class NodeList:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next
        
    def __str__(self) -> str:
        return f"{self.value}"
        
        
def remove_nth_from_end(head: NodeList, n: int) -> NodeList:
    
    if head is None or head.next is None:
        return head
    
    
    scout, prev = head, head
    
    for _ in range(n):
        scout = scout.next
        
    if scout is None or scout.next is None:
        return head.next

    prev.next = prev.next.next
    
    return prev

    
def print_list(head: NodeList):
    current = head
    
    while current is not None:
        print(f"{current.value} => ", end="")    
        current = current.next

    print("null")



def main():
    head = NodeList(1, NodeList(2, NodeList(3, NodeList(4, None))))
    
    
    print_list(head)
    
    remove_nth_from_end(head, 3)
    
    print_list(head)
    
    
if __name__ == '__main__':
    main()
    
    
    

File: Go/Linked List/Remove The Nth Node From The End.go
package sll

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}

	scout, prev := head, head

	for i := 0; i < n; i++ {
		scout = scout.Next
	}

	if scout == nil {
		return head.Next
	}

	for scout.Next != nil {
		scout = scout.Next
		prev = prev.Next
	}

	prev.Next = prev.Next.Next

	return head

}
